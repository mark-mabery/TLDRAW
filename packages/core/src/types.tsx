/* --------------------- Primary -------------------- */

export interface TLPage<T extends TLShape> {
  id: string;
  name?: string;
  childIndex?: number;
  shapes: Record<string, T>;
  bindings: Record<string, TLBinding>;
  backgroundColor?: string;
}

export interface TLPageState {
  id: string;
  brush?: TLBounds;
  pointedId?: string;
  hoveredId?: string;
  editingId?: string;
  editingBindingId?: string;
  boundsRotation?: number;
  currentParentId?: string;
  selectedIds: string[];
  camera: {
    point: number[];
    zoom: number;
  };
}

export interface TLHandle {
  id: string;
  index: number;
  point: number[];
}

export interface TLShape {
  id: string;
  type: string;
  parentId: string;
  childIndex: number;
  name: string;
  point: number[];
  rotation?: number;
  children?: string[];
  handles?: Record<string, TLHandle>;
  isLocked?: boolean;
  isHidden?: boolean;
  isEditing?: boolean;
  isGenerated?: boolean;
  isAspectRatioLocked?: boolean;
}

export type TLShapeUtils<T extends TLShape> = Record<string, TLShapeUtil<T>>;

export interface TLRenderInfo<T extends SVGElement | HTMLElement = any> {
  isEditing: boolean;
  isBinding: boolean;
  isDarkMode: boolean;
  isCurrentParent: boolean;
  ref?: React.RefObject<T>;
  onTextChange?: TLCallbacks["onTextChange"];
  onTextBlur?: TLCallbacks["onTextBlur"];
  onTextFocus?: TLCallbacks["onTextFocus"];
  onTextKeyDown?: TLCallbacks["onTextKeyDown"];
  onTextKeyUp?: TLCallbacks["onTextKeyUp"];
}

export interface TLTool {
  id: string;
  name: string;
}

export interface TLBinding {
  id: string;
  type: string;
  toId: string;
  fromId: string;
}

export interface TLSettings {
  isDebugMode: boolean;
  isDarkMode: boolean;
  isPenMode: boolean;
}

export interface TLTheme {
  brushFill?: string;
  brushStroke?: string;
  selectFill?: string;
  selectStroke?: string;
  background?: string;
  foreground?: string;
}

export type TLWheelEventHandler = (
  info: TLPointerInfo<string>,
  e: React.WheelEvent<Element> | WheelEvent
) => void;
export type TLPinchEventHandler = (
  info: TLPointerInfo<string>,
  e:
    | React.WheelEvent<Element>
    | WheelEvent
    | React.TouchEvent<Element>
    | TouchEvent
) => void;
export type TLPointerEventHandler = (
  info: TLPointerInfo<string>,
  e: React.PointerEvent
) => void;
export type TLCanvasEventHandler = (
  info: TLPointerInfo<"canvas">,
  e: React.PointerEvent
) => void;
export type TLBoundsEventHandler = (
  info: TLPointerInfo<"bounds">,
  e: React.PointerEvent
) => void;
export type TLBoundsHandleEventHandler = (
  info: TLPointerInfo<TLBoundsCorner | TLBoundsEdge | "rotate">,
  e: React.PointerEvent
) => void;

export interface TLCallbacks {
  onChange: (ids: string[]) => void;

  // Camera events
  onPinchStart: TLPinchEventHandler;
  onPinchEnd: TLPinchEventHandler;
  onPinch: TLPinchEventHandler;
  onPan: TLWheelEventHandler;
  onZoom: TLWheelEventHandler;

  // Pointer Events
  onPointerMove: TLPointerEventHandler;
  onPointerUp: TLPointerEventHandler;
  onPointerDown: TLPointerEventHandler;

  // Canvas (background)
  onPointCanvas: TLCanvasEventHandler;
  onDoubleClickCanvas: TLCanvasEventHandler;
  onRightPointCanvas: TLCanvasEventHandler;
  onDragCanvas: TLCanvasEventHandler;
  onReleaseCanvas: TLCanvasEventHandler;

  // Shape
  onPointShape: TLPointerEventHandler;
  onDoubleClickShape: TLPointerEventHandler;
  onRightPointShape: TLPointerEventHandler;
  onDragShape: TLPointerEventHandler;
  onHoverShape: TLPointerEventHandler;
  onUnhoverShape: TLPointerEventHandler;
  onReleaseShape: TLPointerEventHandler;

  // Bounds (bounding box background)
  onPointBounds: TLBoundsEventHandler;
  onDoubleClickBounds: TLBoundsEventHandler;
  onRightPointBounds: TLBoundsEventHandler;
  onDragBounds: TLBoundsEventHandler;
  onHoverBounds: TLBoundsEventHandler;
  onUnhoverBounds: TLBoundsEventHandler;
  onReleaseBounds: TLBoundsEventHandler;

  // Bounds handles (corners, edges)
  onPointBoundsHandle: TLBoundsHandleEventHandler;
  onDoubleClickBoundsHandle: TLBoundsHandleEventHandler;
  onRightPointBoundsHandle: TLBoundsHandleEventHandler;
  onDragBoundsHandle: TLBoundsHandleEventHandler;
  onHoverBoundsHandle: TLBoundsHandleEventHandler;
  onUnhoverBoundsHandle: TLBoundsHandleEventHandler;
  onReleaseBoundsHandle: TLBoundsHandleEventHandler;

  // Handles (ie the handles of a selected arrow)
  onPointHandle: TLPointerEventHandler;
  onDoubleClickHandle: TLPointerEventHandler;
  onRightPointHandle: TLPointerEventHandler;
  onDragHandle: TLPointerEventHandler;
  onHoverHandle: TLPointerEventHandler;
  onUnhoverHandle: TLPointerEventHandler;
  onReleaseHandle: TLPointerEventHandler;

  // Text
  onTextChange: (id: string, text: string) => void;
  onTextBlur: (id: string) => void;
  onTextFocus: (id: string) => void;
  onTextKeyDown: (id: string, key: string) => void;
  onTextKeyUp: (id: string, key: string) => void;

  // Misc
  onBlurEditingShape: () => void;
  onError: (error: Error) => void;
}

export interface TLBounds {
  minX: number;
  minY: number;
  maxX: number;
  maxY: number;
  width: number;
  height: number;
  rotation?: number;
}

export type TLIntersection = {
  didIntersect: boolean;
  message: string;
  points: number[][];
};

export enum TLBoundsEdge {
  Top = "top_edge",
  Right = "right_edge",
  Bottom = "bottom_edge",
  Left = "left_edge",
}

export enum TLBoundsCorner {
  TopLeft = "top_left_corner",
  TopRight = "top_right_corner",
  BottomRight = "bottom_right_corner",
  BottomLeft = "bottom_left_corner",
}

export interface TLPointerInfo<T extends string = string> {
  target: T;
  pointerId: number;
  origin: number[];
  point: number[];
  delta: number[];
  pressure: number;
  shiftKey: boolean;
  ctrlKey: boolean;
  metaKey: boolean;
  altKey: boolean;
}

export interface TLKeyboardInfo {
  origin: number[];
  point: number[];
  key: string;
  keys: string[];
  shiftKey: boolean;
  ctrlKey: boolean;
  metaKey: boolean;
  altKey: boolean;
}

export interface TLTransformInfo<T extends TLShape> {
  type: TLBoundsEdge | TLBoundsCorner;
  initialShape: T;
  scaleX: number;
  scaleY: number;
  transformOrigin: number[];
}

export interface TLBezierCurveSegment {
  start: number[];
  tangentStart: number[];
  normalStart: number[];
  pressureStart: number;
  end: number[];
  tangentEnd: number[];
  normalEnd: number[];
  pressureEnd: number;
}

/* -------------------------------------------------- */
/*                   Shape Utility                    */
/* -------------------------------------------------- */

export abstract class TLShapeUtil<T extends TLShape> {
  boundsCache = new WeakMap<TLShape, TLBounds>();
  isEditableText = false;
  isAspectRatioLocked = false;
  canEdit = false;

  abstract type: T["type"];

  abstract defaultProps: T;

  abstract render(shape: T, info: TLRenderInfo): JSX.Element;

  abstract renderIndicator(shape: T): JSX.Element | null;

  abstract getBounds(shape: T): TLBounds;

  abstract getRotatedBounds(shape: T): TLBounds;

  abstract hitTest(shape: T, point: number[]): boolean;

  abstract hitTestBounds(shape: T, bounds: TLBounds): boolean;

  abstract transform(
    shape: T,
    bounds: TLBounds,
    info: TLTransformInfo<T>
  ): Partial<T>;

  transformSingle(
    shape: T,
    bounds: TLBounds,
    info: TLTransformInfo<T>
  ): Partial<T> {
    return this.transform(shape, bounds, info);
  }

  shouldRender(_prev: T, _next: T): boolean {
    return true;
  }

  shouldDelete(_shape: T): boolean {
    return false;
  }

  getCenter(shape: T): number[] {
    const bounds = this.getBounds(shape);
    return [bounds.width / 2, bounds.height / 2];
  }

  create(props: Partial<T>): T {
    return { ...this.defaultProps, ...props };
  }

  mutate(shape: T, props: Partial<T>): T {
    return { ...shape, ...props };
  }

  updateChildren<K extends TLShape>(
    _shape: T,
    _children: K[]
  ): Partial<K>[] | void {
    return;
  }

  onChildrenChange(_shape: T, _children: TLShape[]): Partial<T> | void {
    return;
  }

  onBindingChange(
    _shape: T,
    _binding: TLBinding,
    _target: TLShape,
    _targetBounds: TLBounds
  ): Partial<T> | void {
    return;
  }

  onHandleChange(
    _shape: T,
    _handle: Partial<T["handles"]>,
    _info: Partial<TLPointerInfo>
  ): Partial<T> | void {
    return;
  }

  onRightPointHandle(
    _shape: T,
    _handle: Partial<T["handles"]>,
    _info: Partial<TLPointerInfo>
  ): Partial<T> | void {
    return;
  }

  onDoubleClickHandle(
    _shape: T,
    _handle: Partial<T["handles"]>,
    _info: Partial<TLPointerInfo>
  ): Partial<T> | void {
    return;
  }

  onSessionComplete(_shape: T): Partial<T> | void {
    return;
  }

  onBoundsReset(_shape: T): Partial<T> | void {
    return;
  }

  onStyleChange(_shape: T): Partial<T> | void {
    return;
  }
}

/* -------------------- Internal -------------------- */

export interface IShapeTreeNode {
  shape: TLShape;
  children?: IShapeTreeNode[];
  isEditing: boolean;
  isBinding: boolean;
  isDarkMode: boolean;
  isCurrentParent: boolean;
}

/* -------------------------------------------------- */
/*                    Utility Types                   */
/* -------------------------------------------------- */

export type MappedByType<T extends { type: string }> = {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [P in T["type"]]: T extends any ? (P extends T["type"] ? T : never) : never;
};

export type RequiredKeys<T> = {
  [K in keyof T]-?: Record<string, unknown> extends Pick<T, K> ? never : K;
}[keyof T];

export {};
