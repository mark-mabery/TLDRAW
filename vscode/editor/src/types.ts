export enum UI_EVENT {
  TLDRAW_UPDATED = 'TLDRAW_UPDATED',
}

export enum EXTENSION_EVENT {
  LOCAL_FILE_UPDATED = 'LOCAL_FILE_UPDATED',
}
