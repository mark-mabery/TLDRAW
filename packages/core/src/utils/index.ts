import { Utils } from './utils'
export { Utils } from './utils'
export { Svg } from './svg'

export default Utils
