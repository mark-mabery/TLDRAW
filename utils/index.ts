import vec from './vec'
import svg from './svg'
export * from './utils'

export { vec, svg }
