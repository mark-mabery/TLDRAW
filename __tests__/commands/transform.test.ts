import { Corner, Edge, RectangleShape, ShapeType } from 'types'
import { rng } from 'utils'
import TestState from '../test-utils'

describe('transform command', () => {
  const tt = new TestState()
  tt.resetDocumentState()
    .createShape(
      {
        type: ShapeType.Rectangle,
        point: [100, 100],
        size: [100, 100],
        childIndex: 1,
      },
      'rect1'
    )
    .createShape(
      {
        type: ShapeType.Rectangle,
        point: [500, 400],
        size: [200, 200],
        childIndex: 2,
      },
      'rect2'
    )
    .clickShape('rect1')
    .clickShape('rect2', { shiftKey: true })
    .save()

  function getSnapInfo() {
    return {
      rect1: {
        point: tt.getShape<RectangleShape>('rect1').point,
        size: tt.getShape<RectangleShape>('rect1').size,
      },
      rect2: {
        point: tt.getShape<RectangleShape>('rect2').point,
        size: tt.getShape<RectangleShape>('rect2').size,
      },
    }
  }

  it('sets up initial bounds', () => {
    expect(tt.selectedIds).toEqual(['rect1', 'rect2'])

    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 100,
      minY: 100,
      maxX: 700,
      maxY: 600,
      width: 600,
      height: 500,
    })
  })

  describe('when transforming from the bottom-right corner', () => {
    it('does command', () => {
      // Restore the saved data state, with the initial selection
      tt.restore()

      // Move the bounds handle
      tt.startClickingBoundsHandle(Corner.BottomRight)
        .movePointerBy([100, 100])
        .stopClickingBounds()

      // Ensure the bounds have been transformed
      expect(tt.state.values.selectedBounds).toMatchObject({
        minX: 100,
        minY: 100,
        maxX: 800,
        maxY: 700,
        width: 700,
        height: 600,
      })

      expect(getSnapInfo()).toMatchSnapshot()
    })

    it('un-does command', () => {
      // Repeat the same actions, but add an undo at the end
      tt.restore()
        .startClickingBoundsHandle(Corner.BottomRight)
        .movePointerBy([100, 100])
        .stopClickingBounds()
        .undo()

      // Expect the bounds to be the initial bounds
      expect(tt.state.values.selectedBounds).toMatchObject({
        minX: 100,
        minY: 100,
        maxX: 700,
        maxY: 600,
        width: 600,
        height: 500,
      })

      expect(getSnapInfo()).toMatchSnapshot()
    })

    it('re-does command', () => {
      // Repeat the same actions but add an undo and a redo at the end
      tt.restore()
        .startClickingBoundsHandle(Corner.BottomRight)
        .movePointerBy([100, 100])
        .stopClickingBounds()
        .undo()
        .redo()

      // Expect the bounds to be the transformed bounds
      expect(tt.state.values.selectedBounds).toMatchObject({
        minX: 100,
        minY: 100,
        maxX: 800,
        maxY: 700,
        width: 700,
        height: 600,
      })

      expect(getSnapInfo()).toMatchSnapshot()
    })
  })

  // From here on, let's assume that the undo and redos work as expected,
  // so let's only test the command's execution.

  it('transforms from the top edge', () => {
    tt.restore()
      .startClickingBoundsHandle(Edge.Top)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 100,
      minY: 200,
      maxX: 700,
      maxY: 600,
      width: 600,
      height: 400,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the right edge', () => {
    tt.restore()
      .startClickingBoundsHandle(Edge.Right)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 100,
      minY: 100,
      maxX: 800,
      maxY: 600,
      width: 700,
      height: 500,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the bottom edge', () => {
    tt.restore()
      .startClickingBoundsHandle(Edge.Bottom)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 100,
      minY: 100,
      maxX: 700,
      maxY: 700,
      width: 600,
      height: 600,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the left edge', () => {
    tt.restore()
      .startClickingBoundsHandle(Edge.Left)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 200,
      minY: 100,
      maxX: 700,
      maxY: 600,
      width: 500,
      height: 500,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the top-left corner', () => {
    tt.restore()
      .startClickingBoundsHandle(Corner.TopLeft)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 200,
      minY: 200,
      maxX: 700,
      maxY: 600,
      width: 500,
      height: 400,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the top-right corner', () => {
    tt.restore()
      .startClickingBoundsHandle(Corner.TopRight)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 100,
      minY: 200,
      maxX: 800,
      maxY: 600,
      width: 700,
      height: 400,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the bottom-right corner', () => {
    tt.restore()
      .startClickingBoundsHandle(Corner.BottomRight)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 100,
      minY: 100,
      maxX: 800,
      maxY: 700,
      width: 700,
      height: 600,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  it('transforms from the bottom-left corner', () => {
    tt.restore()
      .startClickingBoundsHandle(Corner.BottomLeft)
      .movePointerBy([100, 100])
      .stopClickingBounds()

    // Ensure the bounds have been transformed
    expect(tt.state.values.selectedBounds).toMatchObject({
      minX: 200,
      minY: 100,
      maxX: 700,
      maxY: 700,
      width: 500,
      height: 600,
    })

    expect(getSnapInfo()).toMatchSnapshot()
  })

  describe('snapshot tests', () => {
    it('transforms corners', () => {
      const getRandom = rng('transform-tests-random-number-generator')

      for (const corner of Object.values(Corner)) {
        tt.restore()
          .startClickingBoundsHandle(corner)
          .movePointerBy([getRandom() * 200, getRandom() * 200])
          .stopClickingBounds()

        // Ensure the bounds have been transformed
        expect(tt.state.values.selectedBounds).toMatchSnapshot()

        expect(getSnapInfo()).toMatchSnapshot()
      }
    })

    it('transforms edges', () => {
      const getRandom = rng('transform-tests-random-number-generator')

      for (const edge of Object.values(Edge)) {
        tt.restore()
          .startClickingBoundsHandle(edge)
          .movePointerBy([getRandom() * 200, getRandom() * 200])
          .stopClickingBounds()

        // Ensure the bounds have been transformed
        expect(tt.state.values.selectedBounds).toMatchSnapshot()

        expect(getSnapInfo()).toMatchSnapshot()
      }
    })

    it('shift-transforms corners', () => {
      const getRandom = rng('transform-tests-random-number-generator')

      for (const corner of Object.values(Corner)) {
        tt.restore()
          .startClickingBoundsHandle(corner)
          .movePointerBy([getRandom() * 200, getRandom() * 200], {
            shiftKey: true,
          })
          .stopClickingBounds()

        // Ensure the bounds have been transformed
        expect(tt.state.values.selectedBounds).toMatchSnapshot()

        expect(getSnapInfo()).toMatchSnapshot()
      }
    })

    it('shift-transforms edges', () => {
      const getRandom = rng('transform-tests-random-number-generator')

      for (const edge of Object.values(Edge)) {
        tt.restore()
          .startClickingBoundsHandle(edge)
          .movePointerBy([getRandom() * 200, getRandom() * 200], {
            shiftKey: true,
          })
          .stopClickingBounds()

        // Ensure the bounds have been transformed
        expect(tt.state.values.selectedBounds).toMatchSnapshot()

        expect(getSnapInfo()).toMatchSnapshot()
      }
    })
  })
})
