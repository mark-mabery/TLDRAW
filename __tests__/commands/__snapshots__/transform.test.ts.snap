// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform command snapshot tests shift-transforms corners 1`] = `
Object {
  "height": 593.73,
  "maxX": 700,
  "maxY": 600,
  "minX": -12.476,
  "minY": 6.27,
  "width": 712.476,
}
`;

exports[`transform command snapshot tests shift-transforms corners 2`] = `
Object {
  "rect1": Object {
    "point": Array [
      -12.476,
      6.27,
    ],
    "size": Array [
      118.75,
      118.75,
    ],
  },
  "rect2": Object {
    "point": Array [
      462.51,
      362.51,
    ],
    "size": Array [
      237.49,
      237.49,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms corners 3`] = `
Object {
  "height": 531.6320000000001,
  "maxX": 737.9599999999999,
  "maxY": 600,
  "minX": 100,
  "minY": 68.368,
  "width": 637.9599999999999,
}
`;

exports[`transform command snapshot tests shift-transforms corners 4`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      68.368,
    ],
    "size": Array [
      106.33,
      106.33,
    ],
  },
  "rect2": Object {
    "point": Array [
      525.31,
      387.35,
    ],
    "size": Array [
      212.65,
      212.65,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms corners 5`] = `
Object {
  "height": 533.62,
  "maxX": 740.3499999999999,
  "maxY": 633.62,
  "minX": 100,
  "minY": 100,
  "width": 640.3499999999999,
}
`;

exports[`transform command snapshot tests shift-transforms corners 6`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      106.72,
      106.72,
    ],
  },
  "rect2": Object {
    "point": Array [
      526.9,
      420.17,
    ],
    "size": Array [
      213.45,
      213.45,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms corners 7`] = `
Object {
  "height": 490.52,
  "maxX": 700,
  "maxY": 590.52,
  "minX": 111.38,
  "minY": 100,
  "width": 588.62,
}
`;

exports[`transform command snapshot tests shift-transforms corners 8`] = `
Object {
  "rect1": Object {
    "point": Array [
      111.38,
      100,
    ],
    "size": Array [
      98.104,
      98.104,
    ],
  },
  "rect2": Object {
    "point": Array [
      503.79,
      394.31,
    ],
    "size": Array [
      196.21,
      196.21,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms edges 1`] = `
Object {
  "height": 593.73,
  "maxX": 756.24,
  "maxY": 600,
  "minX": 43.762,
  "minY": 6.27,
  "width": 712.4780000000001,
}
`;

exports[`transform command snapshot tests shift-transforms edges 2`] = `
Object {
  "rect1": Object {
    "point": Array [
      43.762,
      6.27,
    ],
    "size": Array [
      118.75,
      118.75,
    ],
  },
  "rect2": Object {
    "point": Array [
      518.75,
      362.51,
    ],
    "size": Array [
      237.49,
      237.49,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms edges 3`] = `
Object {
  "height": 531.6260000000001,
  "maxX": 737.9599999999999,
  "maxY": 615.8100000000001,
  "minX": 100,
  "minY": 84.184,
  "width": 637.9599999999999,
}
`;

exports[`transform command snapshot tests shift-transforms edges 4`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      84.184,
    ],
    "size": Array [
      106.33,
      106.33,
    ],
  },
  "rect2": Object {
    "point": Array [
      525.31,
      403.16,
    ],
    "size": Array [
      212.65,
      212.65,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms edges 5`] = `
Object {
  "height": 411.35,
  "maxX": 646.81,
  "maxY": 511.35,
  "minX": 153.19,
  "minY": 100,
  "width": 493.61999999999995,
}
`;

exports[`transform command snapshot tests shift-transforms edges 6`] = `
Object {
  "rect1": Object {
    "point": Array [
      153.19,
      100,
    ],
    "size": Array [
      82.269,
      82.269,
    ],
  },
  "rect2": Object {
    "point": Array [
      482.27,
      346.81,
    ],
    "size": Array [
      164.54,
      164.54,
    ],
  },
}
`;

exports[`transform command snapshot tests shift-transforms edges 7`] = `
Object {
  "height": 490.52,
  "maxX": 700,
  "maxY": 595.26,
  "minX": 111.38,
  "minY": 104.74,
  "width": 588.62,
}
`;

exports[`transform command snapshot tests shift-transforms edges 8`] = `
Object {
  "rect1": Object {
    "point": Array [
      111.38,
      104.74,
    ],
    "size": Array [
      98.104,
      98.104,
    ],
  },
  "rect2": Object {
    "point": Array [
      503.79,
      399.05,
    ],
    "size": Array [
      196.21,
      196.21,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms corners 1`] = `
Object {
  "height": 593.73,
  "maxX": 700,
  "maxY": 600,
  "minX": 27.892,
  "minY": 6.27,
  "width": 672.108,
}
`;

exports[`transform command snapshot tests transforms corners 2`] = `
Object {
  "rect1": Object {
    "point": Array [
      27.892,
      6.27,
    ],
    "size": Array [
      112.02,
      118.75,
    ],
  },
  "rect2": Object {
    "point": Array [
      475.96,
      362.51,
    ],
    "size": Array [
      224.04,
      237.49,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms corners 3`] = `
Object {
  "height": 490.17,
  "maxX": 737.9599999999999,
  "maxY": 600,
  "minX": 100,
  "minY": 109.83,
  "width": 637.9599999999999,
}
`;

exports[`transform command snapshot tests transforms corners 4`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      109.83,
    ],
    "size": Array [
      106.33,
      98.034,
    ],
  },
  "rect2": Object {
    "point": Array [
      525.31,
      403.93,
    ],
    "size": Array [
      212.65,
      196.07,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms corners 5`] = `
Object {
  "height": 411.35,
  "maxX": 740.3499999999999,
  "maxY": 511.35,
  "minX": 100,
  "minY": 100,
  "width": 640.3499999999999,
}
`;

exports[`transform command snapshot tests transforms corners 6`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      106.72,
      82.269,
    ],
  },
  "rect2": Object {
    "point": Array [
      526.9,
      346.81,
    ],
    "size": Array [
      213.45,
      164.54,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms corners 7`] = `
Object {
  "height": 437.4,
  "maxX": 700,
  "maxY": 537.4,
  "minX": 111.38,
  "minY": 100,
  "width": 588.62,
}
`;

exports[`transform command snapshot tests transforms corners 8`] = `
Object {
  "rect1": Object {
    "point": Array [
      111.38,
      100,
    ],
    "size": Array [
      98.104,
      87.479,
    ],
  },
  "rect2": Object {
    "point": Array [
      503.79,
      362.44,
    ],
    "size": Array [
      196.21,
      174.96,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms edges 1`] = `
Object {
  "height": 593.73,
  "maxX": 700,
  "maxY": 600,
  "minX": 100,
  "minY": 6.27,
  "width": 600,
}
`;

exports[`transform command snapshot tests transforms edges 2`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      6.27,
    ],
    "size": Array [
      100,
      118.75,
    ],
  },
  "rect2": Object {
    "point": Array [
      500,
      362.51,
    ],
    "size": Array [
      200,
      237.49,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms edges 3`] = `
Object {
  "height": 500,
  "maxX": 737.9599999999999,
  "maxY": 600,
  "minX": 100,
  "minY": 100,
  "width": 637.9599999999999,
}
`;

exports[`transform command snapshot tests transforms edges 4`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      106.33,
      100,
    ],
  },
  "rect2": Object {
    "point": Array [
      525.31,
      400,
    ],
    "size": Array [
      212.65,
      200,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms edges 5`] = `
Object {
  "height": 411.35,
  "maxX": 700,
  "maxY": 511.35,
  "minX": 100,
  "minY": 100,
  "width": 600,
}
`;

exports[`transform command snapshot tests transforms edges 6`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      100,
      82.269,
    ],
  },
  "rect2": Object {
    "point": Array [
      500,
      346.81,
    ],
    "size": Array [
      200,
      164.54,
    ],
  },
}
`;

exports[`transform command snapshot tests transforms edges 7`] = `
Object {
  "height": 500,
  "maxX": 700,
  "maxY": 600,
  "minX": 111.38,
  "minY": 100,
  "width": 588.62,
}
`;

exports[`transform command snapshot tests transforms edges 8`] = `
Object {
  "rect1": Object {
    "point": Array [
      111.38,
      100,
    ],
    "size": Array [
      98.104,
      100,
    ],
  },
  "rect2": Object {
    "point": Array [
      503.79,
      400,
    ],
    "size": Array [
      196.21,
      200,
    ],
  },
}
`;

exports[`transform command transforms from the bottom edge 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      100,
      120,
    ],
  },
  "rect2": Object {
    "point": Array [
      500,
      460,
    ],
    "size": Array [
      200,
      240,
    ],
  },
}
`;

exports[`transform command transforms from the bottom-left corner 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      200,
      100,
    ],
    "size": Array [
      83.333,
      120,
    ],
  },
  "rect2": Object {
    "point": Array [
      533.33,
      460,
    ],
    "size": Array [
      166.67,
      240,
    ],
  },
}
`;

exports[`transform command transforms from the bottom-right corner 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      116.67,
      120,
    ],
  },
  "rect2": Object {
    "point": Array [
      566.67,
      460,
    ],
    "size": Array [
      233.33,
      240,
    ],
  },
}
`;

exports[`transform command transforms from the left edge 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      200,
      100,
    ],
    "size": Array [
      83.333,
      100,
    ],
  },
  "rect2": Object {
    "point": Array [
      533.33,
      400,
    ],
    "size": Array [
      166.67,
      200,
    ],
  },
}
`;

exports[`transform command transforms from the right edge 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      116.67,
      100,
    ],
  },
  "rect2": Object {
    "point": Array [
      566.67,
      400,
    ],
    "size": Array [
      233.33,
      200,
    ],
  },
}
`;

exports[`transform command transforms from the top edge 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      200,
    ],
    "size": Array [
      100,
      80,
    ],
  },
  "rect2": Object {
    "point": Array [
      500,
      440,
    ],
    "size": Array [
      200,
      160,
    ],
  },
}
`;

exports[`transform command transforms from the top-left corner 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      200,
      200,
    ],
    "size": Array [
      83.333,
      80,
    ],
  },
  "rect2": Object {
    "point": Array [
      533.33,
      440,
    ],
    "size": Array [
      166.67,
      160,
    ],
  },
}
`;

exports[`transform command transforms from the top-right corner 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      200,
    ],
    "size": Array [
      116.67,
      80,
    ],
  },
  "rect2": Object {
    "point": Array [
      566.67,
      440,
    ],
    "size": Array [
      233.33,
      160,
    ],
  },
}
`;

exports[`transform command when transforming from the bottom-right corner does command 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      116.67,
      120,
    ],
  },
  "rect2": Object {
    "point": Array [
      566.67,
      460,
    ],
    "size": Array [
      233.33,
      240,
    ],
  },
}
`;

exports[`transform command when transforming from the bottom-right corner re-does command 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      116.67,
      120,
    ],
  },
  "rect2": Object {
    "point": Array [
      566.67,
      460,
    ],
    "size": Array [
      233.33,
      240,
    ],
  },
}
`;

exports[`transform command when transforming from the bottom-right corner un-does command 1`] = `
Object {
  "rect1": Object {
    "point": Array [
      100,
      100,
    ],
    "size": Array [
      100,
      100,
    ],
  },
  "rect2": Object {
    "point": Array [
      500,
      400,
    ],
    "size": Array [
      200,
      200,
    ],
  },
}
`;
